protocol/streaming/grpc/grpc_streaming_manager.go

```diff
@@ -36,16 +53,78 @@ type OrderbookSubscription struct {
 	srv clobtypes.Query_StreamOrderbookUpdatesServer
 }
 
-func NewGrpcStreamingManager() *GrpcStreamingManagerImpl {
-	return &GrpcStreamingManagerImpl{
+func NewGrpcStreamingManager(
+	logger log.Logger,
+	bufferWindow uint32,
+) *GrpcStreamingManagerImpl {
+	grpcStreamingManager := &GrpcStreamingManagerImpl{
+		logger:                 logger.With("module", "grpc-streaming"),
 		orderbookSubscriptions: make(map[uint32]*OrderbookSubscription),
+		nextSubscriptionId:     0,
+
+		updateBuffer:           make(chan bufferInternalResponse, bufferWindow),
+		updateBufferWindowSize: bufferWindow,
 	}
+
+	// Worker goroutine to consistently read from channel and send out updates
+	go func() {
+		for internalResponse := range grpcStreamingManager.updateBuffer {
+			grpcStreamingManager.sendUpdateResponse(internalResponse)
+		}
+	}()
+
+	return grpcStreamingManager
 }
 
 func (sm *GrpcStreamingManagerImpl) Enabled() bool {
 	return true
 }
 
+func (sm *GrpcStreamingManagerImpl) Stop() {
+	close(sm.updateBuffer)
+}
+
+func (sm *GrpcStreamingManagerImpl) EmitMetrics() {
+	metrics.SetGauge(metrics.GrpcStreamingBufferSize, float32(len(sm.updateBuffer)))
+	metrics.SetGauge(metrics.GrpcStreamingNumConnections, float32(len(sm.orderbookSubscriptions)))
+}
+
+func (sm *GrpcStreamingManagerImpl) removeSubscription(id uint32) {
+	sm.logger.Info(
+		fmt.Sprintf(
+			"Removing orderbook subscriptions for subscription id %+v",
+			id,
+		),
+	)
+	delete(sm.orderbookSubscriptions, id)
+}
+
+func (sm *GrpcStreamingManagerImpl) sendUpdateResponse(
+	internalResponse bufferInternalResponse,
+) {
+	// Send update to subscribers.
+	subscriptionIdsToRemove := make([]uint32, 0)
+
+	for id, subscription := range sm.orderbookSubscriptions {
```

Comment:
do we need locking here? or sync.Map?

Reply:
discussed in person. non blocking
https://linear.app/dydx/issue/CT-869/grpc-streaming-correctness-investigate-locking-or-syncmap-for